### PR TITLE
Working version for 2.0.0

### PR DESCRIPTION
**Pull request by bvogel**
Hi,

this version is working for me for our use cases (filling a pre-existing PDF, concatenating 2 PDFs), working in the sense that the result looks correct. However half of the specs are failing and I'm a bit at a loss whether that's something fundamentally flawed or I would have to adapt the specs to the current code. 
I would like to hear from others whether the changes will work for them as well.

### CODE REVIEW

**Comment by packetmonkey** [comment]
I know that the original implementation of templates was considered fundamentally flawed where you could easily cause conflicts when we merged the PDF trees.

I haven't looked at this code probably ever and you are the first person to show interest in it, so congratulations for that :)

Lets leave this open a bit to see if we get any thoughts. I know I would prefer a passing test suite which may involve refactoring to make everything work well.

Also can you please not bump the version number? We will do that if/when we go to release a new version of the gem.

Thanks for your help! If I get a chance I'll look at the specs as well, but most of my time in the near term will be trying to get ready for a release of prawn this weekend so I'm not sure when I will get a chance to check it out.

**Comment by bvogel** [comment]
Anyone has tried this? Anyone with some input regarding the specs?

**Comment by bvogel** [comment]
Hi again (@packetmonkey or @practicingruby),

I had another go at the template code and I'm now positive it will work in a broad range of use-cases. One issue remains however where I#m completely stumped: Fonts don't seem to be added from the template to the resulting document, causing two specs to fail. This automagically worked previously, but I would need pointers as to where start looking, why they don't get added currently. Thanks a bunch for the continued support!

B.

**Comment by gettalong** [comment]
Another thing you might want to look at: The contents of a page is defined by the associated content streams. However, since it is not possible to refer to indirect PDF object in the content stream, one has to use named resources. For example, an image is drawn like this: `/Image1 Do`.

These resources are defined in the /Resources dictionary of a PDF page object (see PDF1.7 s7.8 and especially s7.8.3). Not only are images referenced via the /Resources dictionary but also fonts.

So what you could test is if importing a template with a picture works. If it doesn't, then the /Resources dictionary and all referenced resources are probably not correctly copied to the target document. If it does, you might want to look at that code and see why images (defined in the /XObject sub-dictionary of /Resources) are handled correctly and why fonts are not.

**Comment by bvogel** [comment]
Hi @gettalong, well out current use case has a template with images, fonts and such and that works flawlessly, images and fonts are copied - they render ok in Adobe and I can see them in the tree in iText-rups (a wonderful tool btw). What I can't figure out is whether that is coincidence for our template or the current spec is inadequate. I doubt the latter as the PDF generated with the spec definitely looks off and is missing the font in the /Resources/Font path, however our PDF with embedded fonts and images has them included only difference would be that our PDF specifies version 1.4, whether the spec doesn't specify a version.

**Comment by gettalong** [comment]
Can you point me at the spec that is failing?

**Comment by bvogel** [comment]
These two are failing:
rspec ./spec/template_spec.rb:146 # Document built from a template should correctly add a TTF font to a template that has existing fonts
rspec ./spec/template_spec.rb:298 # Document#start_new_page with :template option should correctly add a TTF font to a template that has existing fonts

**Comment by gettalong** [comment]
Thanks @bvogel ! The problem seems to come from Prawn not verifying that the name for the font is not already used! I.e. there are two font resources named `/F1.0` and therefore the first example doesn't work. After manually changing `/F1.0` to `/F2.0` in the template PDF, the spec passes.

So this is actually a prawn or pdf-core problem.

**Comment by bvogel** [comment]
Good Find! Thanks. I actually saw that piece of code that checked whether a Font with the same name already existed and would skip over the routine that would add the font. This behavior is correct for regular prawn use cases as prawn will add all fonts. With a template a font is added without prawn knowing about it, so it is an error related to templates. 
Would the correct way to fix this issue IMO be to fill the font-registry with the template fonts? As I see it the duplication of the unique Id is caused by prawn because it thinks the font-registry is empty?

**Comment by gettalong** [comment]
I probably wouldn't register the template fonts but modify the `#generate_unique_id` (https://github.com/prawnpdf/prawn/blob/master/lib/prawn/font.rb#L398) method to really generate a unique id.

Nota bene: This was just from a 5min look at the code, I'm not really familiar with it.

**Comment by bvogel** [comment]
squashed the changes to be one sole commit, when PR https://github.com/prawnpdf/prawn/pull/898 has been accepted all specs will pass

**Comment by crazymykl** [comment]
Needs a rebase.

**Comment by bvogel** [comment]
I know...
I'm on it.
Sent via bush  drums
Am 08.08.2015 17:06 schrieb Mike MacDonald <notifications@github.com>:Needs a rebase.

File: lib/prawn/templates.rb [hunk 1]
```diff
@@ -246,4 +258,4 @@ def load_object_graph(hash, object)
 Prawn::Document::VALID_OPTIONS << :template
 Prawn::Document.extensions << Prawn::Templates
 
-PDF::Core::ObjectStore.send(:include, Prawn::Templates::ObjectStoreExtensions)
```
Comment (crazymykl) [inline lib/prawn/templates.rb]: This change is breaking CI.

Comment (bvogel) [inline lib/prawn/templates.rb]: I  see, actually I previously had support for ruby 2.0.0+ only. Should I try to maintain 1.9.3 until praw drops it?

**Comment by gettalong** [comment]
One thing to consider: You add files named after the original files in pdf-core and prawn. Personally, I wouldn't do that because it *may* become a problem (e.g. if prawn-template was required first).

I would create two files, `lib/pdf/core/prawn_template_extensions.rb` and `lib/prawn/prawn_template_extensions.rb` and put all the code added to pdf-core and prawn there. The main entry point `lib/prawn/templates.rb` should then require the extension files and modify prawn so that templates are available.

Using `require_relative` might work in this case but having files with the same name *might* become a problem.

Just my 2c.

**Comment by bvogel** [comment]
Hi Thomas (@gettalong),
I thought about it and even stumbled onto something related (see prawn/text.rb vs pdf/core/text.rb), I'll look into that for the next release. I'd like to get this in the hand of people so we'll get more feedback. I'm even thinking about asking to open the issue tracker for this. But currently I want to refrain from redoing all the code again. It helps in that sense that some methods are just copies of the current code on prawn master with the template related parts re-added. This makes it easier to track changes there and follow changes here.

**Comment by gettalong** [comment]
I understand your intentions, I just wanted to point out that this might come back and bite us.

Thanks for you work on this!

**Comment by bvogel** [comment]
Who should merge this? Evan or Gregory?

**Comment by gettalong** [comment]
I would wait until prawn was released with the prawnpdf/prawn#898 patch, then change the needed version number in the gemspec and if travis gives the okay, I can push the Github merge button or you can do it yourself.

This repo is currently unmaintained so it should be okay if you merge this yourself.

**Comment by practicingruby** [comment]
@bvogel: You're the one investing the most time and energy into this, so its fine for you to consider yourself the acting maintainer of this repository, even if you don't want to make that an ongoing commitment.

**Comment by packetmonkey** [comment]
@bvogel I went ahead and opened up issues and the wiki on the repo for you.

Like @practicingruby said, you are putting the most work into this code base, if you want go ahead and take point on merging whatever you like and you can work with the prawn core team so we can add any features you need to make this code base work as well as possible.

I have similar concerns to @gettalong with regards to the code duplication but as long as you are aware, at least you are looking at the code where before it was totally abandoned. I suggest we leave large warnings in the README about known issues and possible compatibility problems. 

Otherwise we can merge all the PRs in and clean up what's left then maybe get the gem ready for another release.

Sound good?

**Comment by bvogel** [comment]
Hi guys, 

Ok, I'll put myself into the drivers seat. I'll wait for https://github.com/prawnpdf/prawn/pull/898 to be merged and released, as I prefer to release on green CI, as well I'd raise the version a minor version (0.1.0) to mark a breaking change. In short, I'll take it from here. 

Thanks guys.

**Comment by gettalong** [comment]
:+1:

**Comment by stayhero** [comment]
:+1: 

I'm also interested in seeing the template stuff maintained again. We're using it, and... from my point of view exactly the possibility to open and edit existing PDFs is what made Prawn stand out against other possible ways to generate PDFs. :-)

**Comment by bvogel** [comment]
@stayhero would you have the chance to try this out? Just point prawn and prawn-templates to our repos...

**Comment by stayhero** [comment]
@bvogel  I've tried today and got an error (some method called ...stream missing). I tried with latest prawn and prawn-templates. 

You say I should try with your prawn fork (bvogel/prawn.git) and this prawn-templates gem (prawnpdf/prawn-templates)?

**Comment by stayhero** [comment]
Er... ok, got it. I should use ReverseRetail/prawn-templates and ReverseRetail/prawn, right? Will check this out today.

**Comment by bvogel** [comment]
@stayhero - not quite: please use my fork of prawn (`bvogel/prawn`) and our companies fork of prawn-templates (`ReverseRetail/prawn-templates`)

**Comment by practicingruby** [comment]
@bvogel: In the future, it may be easiest to create branches directly on the `prawnpdf/prawn-templates` repository... this makes them revisable by all Prawn contributors, and makes it so downstream users need to set up fewer remotes.

**Comment by stayhero** [comment]
@bvogel :+1:  Works with your fork and your companies prawn-templates gem for us.

**Comment by HannesBenson** [comment]
@bvogel :+1: This resolved the issues I was experiencing as well.

**Comment by ollie** [comment]
Thanks @bvogel worked for us too. Would love to see it merged in upstream(s). :)

**Comment by jpalumickas** [comment]
Thanks @bvogel ! When we can expect this merged into master :wink: ?

**Comment by bvogel** [comment]
@jpalumickas I have to wait for a prawn 2.0.3 as I need prawnpdf/prawn#898 to be included in a release. Once thats through I will publish this as a release.

**Comment by jpalumickas** [comment]
Thanks for information !

**Pull request merged**